Weapon Parsing & DPS Tooltip on Item Info

berserkingyadis opened the pull request:
Adding support for weapon parsing and adding DPS tooltip (map info equivalent)

This is work in progress, im learning the project structure

berserkingyadis commented:
This is now in a discussable state. Please review.

**What I added:**
- Weapon damage and attack speed parsing and DPS calculation
- A new InfoListener for Weapons that displays the Weapon Name, Icon and DPS
- Added more stats to the WeaponItem class for damage values

**What I fixed:**
- Rune Daggers are recognized as weapons now

**What it looks like:**
ALT-A on a few of the weapons i have found:

![PathOfExile_x64_aIjK3rCPdr](https://user-images.githubusercontent.com/9959527/72889651-76324b00-3d10-11ea-8f65-f6643066eca0.png)
![PathOfExile_x64_sKfmR0b5JP](https://user-images.githubusercontent.com/9959527/72889657-79c5d200-3d10-11ea-9672-0bc31c7fdbd5.png)
![PathOfExile_x64_gDA0cc6G3Q](https://user-images.githubusercontent.com/9959527/72889663-7b8f9580-3d10-11ea-9c86-fd1dd660daa2.png)

**What is not yet working correctly:**
- Elemental Damage can be Cold, Lightning or Fire in the Stats part of the weapon and I just add them all together as "total Elemental DPS". I have not found a deterministic way to tell which numbers belong to which element in the Weapon stat text.
```
Elemental Damage: 77-140 (augmented), 62-110 (augmented), 1-237 (augmented)
```

pvaass commented on src/main/java/dev/tricht/lunaris/item/ItemParser.java:
```diff
@@ -32,6 +33,47 @@ public Item parse() {
         }
 
+        //If the item is a weapon, parse damage and attackspeed stats for DPS calculation
+        if (itemType instanceof WeaponItem){
```
Should probably move this to StatsPart (or make something new) to keep the ItemParser as simple as possible

berserkingyadis replied:
Ill look into moving this logic into the StatsPart class

pvaass commented on src/main/java/dev/tricht/lunaris/item/ItemParser.java:
```diff
@@ -32,6 +33,47 @@ public Item parse() {
+                else if(stat.contains("Elemental Damage")){
+                    ((WeaponItem) itemType).setFireMin(digits.get(0));
+                    ((WeaponItem) itemType).setFireMax(digits.get(1));
+                    if(digits.size()>2){
```
If I understand correctly, if an item has Fire and Lightning damage, this code will confuse the Lightning damage for Cold damage?

I think to fix this we could check the mods instead of the stats (unless there are items with innate elemental damage?)
Will be considerably more difficult to implement though

berserkingyadis replied:
If we don't want to differentiate between the different types of elemental damage and just display "Elemental DPS" then this is fine IMO.

If we want to display Phys+Elem+Fire+Lightning+Cold+Chaos+Total then more has to be done here. Yea I could get the Damage Elements from the Mod part, this wont really be more difficult it will a bit more work tho.

Do you want it to be more detailed?

pvaass commented on src/main/java/dev/tricht/lunaris/item/parser/StatsPart.java:
```diff
@@ -31,7 +31,7 @@ public ItemType getWeaponType() {
             return new WeaponItem(WeaponType.CLAW);
         }
-        if (name.equals("Dagger")) {
+        if (name.equals("Dagger") || name.equals("Rune Dagger")) {
```
Nice one

pvaass commented on src/main/java/dev/tricht/lunaris/listeners/WeaponInfoListener.java:
```diff
@@ -0,0 +1,57 @@
+
+        int row = 1;
+        elements.put(new dev.tricht.lunaris.elements.Label("Total dps: " + df.format(totalDPS)), new int[]{1, row++});
+        if(physDPS!=0)elements.put(new dev.tricht.lunaris.elements.Label("PhysDPS: "+ df.format(physDPS)), new int[]{1, row++});
```
We could probably just write this in full (Physical DPS, Elemental DPS, Chaos DPS), it's not like we're lacking screen space or anything

berserkingyadis replied:
Will do. Also if we choose to add the single Elements to this, the Tooltip will get bigger.

berserkingyadis commented:
@pvaass 

I changed the DPS Overlay and refactored the DPS parsing & calculation into the Stats part. 

The only unresolved point ATM is the more detailed display of the Elemental DPS which IMO is not important for a first release of this feature.

When there is nothing else to change in this PR my vote is to merge it into master.

![PathOfExile_x64_OhXRv62iJA](https://user-images.githubusercontent.com/9959527/73121513-02e04180-3f7b-11ea-9c92-22ddda495abf.png)
![PathOfExile_x64_4pD3iOpeKN](https://user-images.githubusercontent.com/9959527/73121515-05db3200-3f7b-11ea-9120-25fd9226074a.png)

pvaass commented:
Merged, thanks a lot!
